gtsfm/utils/homography_decomposition.py

```diff
@@ -0,0 +1,387 @@
+    if points1.shape != points2.shape:
+        raise RuntimeError("Coordinates of 2d correspondences must have the same shape.")
+
+    K1 = camera_intrinsics_i1.K()
+    K2 = camera_intrinsics_i2.K()
+    i2Ri1_cmbs, i2ti1_cmbs, n_cmbs = decompose_homography_matrix(H, K1, K2)
+
+    points3D = []
+    for i in range(len(i2Ri1_cmbs)):
+        points3D_cmb = check_cheirality(
+            i2Ri1_cmbs[i], i2ti1_cmbs[i], camera_intrinsics_i1, camera_intrinsics_i2, points1, points2
+        )
+        print(f"hypothesis {i}  -> {len(points3D_cmb)}")
+        print(i2ti1_cmbs[i])
+        if len(points3D_cmb) >= len(points3D):
+            i2Ri1 = i2Ri1_cmbs[i]
+            i2ti1 = i2ti1_cmbs[i]
+            n = n_cmbs[i]
+            points3D = points3D_cmb
+
+    points3D = np.array(points3D)
+
+    # Note: if norm(i2ti1) == 0, this is a pure rotation. Otherwise, planar scene.
+
+    # Note: we cannot blindly cast the translation direction to Unit3, as zero translation is valid
+    # from a rotation, and should be noted in 1dsfm.
+    return i2Ri1, Unit3(i2ti1), n, points3D
+
+
+def check_cheirality(
+    i2Ri1: Rot3,
+    i2ti1: np.ndarray,
+    camera_intrinsics_i1: Cal3Bundler,
+    camera_intrinsics_i2: Cal3Bundler,
+    points1: np.ndarray,
+    points2: np.ndarray,
+) -> np.ndarray:
+    """
+    Args:
+        R: array of shape (3,3)
+        t: array of shape (3,)
+        points1:
+        points2:
+
+    Returns:
+        points3D: array of shape (N,3)
+    """
+    if points1.shape != points2.shape:
+        raise RuntimeError("Coordinates of 2d correspondences must have the same shape.")
+
+    i2Ti1 = Pose3(i2Ri1, i2ti1)
+    camera_dict = {
+        1: PinholeCameraCal3Bundler(i2Ti1, camera_intrinsics_i1),
+        2: PinholeCameraCal3Bundler(Pose3(), camera_intrinsics_i2),
+    }
+
+    # set the camera 2 pose as the global coordinate system
+    # COLMAP states that wTc for camera 2 is equal to (R,t) from homography.
+    # given wTi1 and wTi2, then if
+    # i2Ti1 = wTi2.inverse() * wTi1
+    # thus, wTi2 is identity, and wTi1 = i2Ti1
+    triangulator = Point3dInitializer(
+        track_camera_dict=camera_dict, mode=TriangulationParam.NO_RANSAC, reproj_error_thresh=float("inf")
+    )
+
+    points3d = []
+    # try triangulating each point
+    for point1, point2 in zip(points1, points2):
+
+        track_2d = SfmTrack2d(measurements=[SfmMeasurement(i=1, uv=point1), SfmMeasurement(i=2, uv=point2)])
+        track_3d, _, exit_code = triangulator.triangulate(track_2d)
+        if exit_code == TriangulationExitCode.CHEIRALITY_FAILURE:
+            continue
+        points3d.append(track_3d.point3())
+
+    return points3d
+
+
+def decompose_homography_matrix(
+    H: np.ndarray, K1: np.ndarray, K2: np.ndarray
+) -> Tuple[List[Rot3], List[Unit3], List[Unit3]]:
+    """Decompose an homography matrix into the possible rotations, translations, and plane normal vectors.
+
+    Based off of the OpenCV and COLMAP implementations:
```

Comment:
I see so many "based off of"s and it seems more informal (and a bit weird) to me, but maybe this is a personal choice, so feel free to ignore the other comments if you prefer.  
https://english.stackexchange.com/questions/15081/based-on-instead-of-based-off-of :smile: